freqtrade/rpc/external_message_consumer.py

```diff
@@ -0,0 +1,341 @@
+                                                 WSWhitelistRequest)
+
+
+if TYPE_CHECKING:
+    import websockets.connect
+    import websockets.exceptions
+
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalMessageConsumer:
+    """
+    The main controller class for consuming external messages from
+    other FreqTrade bot's
+    """
+
+    def __init__(
+        self,
+        config: Dict[str, Any],
+        dataprovider: DataProvider
+    ):
+        self._config = config
+        self._dp = dataprovider
+
+        self._running = False
+        self._thread = None
+        self._loop = None
+        self._main_task = None
+        self._sub_tasks = None
+
+        self._emc_config = self._config.get('external_message_consumer', {})
+
+        self.enabled = self._emc_config.get('enabled', False)
+        self.producers = self._emc_config.get('producers', [])
+
+        self.wait_timeout = self._emc_config.get('wait_timeout', 300)  # in seconds
+        self.ping_timeout = self._emc_config.get('ping_timeout', 10)  # in seconds
+        self.sleep_time = self._emc_config.get('sleep_time', 10)  # in seconds
+
+        # The amount of candles per dataframe on the initial request
+        self.initial_candle_limit = self._emc_config.get('initial_candle_limit', 1500)
+
+        # Message size limit, in megabytes. Default 8mb, Use bitwise operator << 20 to convert
+        # as the websockets client expects bytes.
+        self.message_size_limit = (self._emc_config.get('message_size_limit', 8) << 20)
+
+        self.validate_config()
+
+        # Setting these explicitly as they probably shouldn't be changed by a user
+        # Unless we somehow integrate this with the strategy to allow creating
+        # callbacks for the messages
+        self.topics = [RPCMessageType.WHITELIST, RPCMessageType.ANALYZED_DF]
+
+        # Allow setting data for each initial request
+        self._initial_requests: List[WSRequestSchema] = [
+            WSSubscribeRequest(data=self.topics),
+            WSWhitelistRequest(),
+            WSAnalyzedDFRequest()
+        ]
+
+        # Specify which function to use for which RPCMessageType
+        self._message_handlers: Dict[str, Callable[[str, WSMessageSchema], None]] = {
+            RPCMessageType.WHITELIST: self._consume_whitelist_message,
+            RPCMessageType.ANALYZED_DF: self._consume_analyzed_df_message,
+        }
+
+        self.start()
+
+    def validate_config(self):
+        """
+        Make sure values are what they are supposed to be
+        """
+        if self.enabled and len(self.producers) < 1:
+            raise OperationalException("You must specify at least 1 Producer to connect to.")
+
+        if self.enabled and self._config.get('process_only_new_candles', True):
+            # Warning here or require it?
+            logger.warning("To receive best performance with external data,"
+                           "please set `process_only_new_candles` to False")
```

Comment:
Leave it as a warning for now.
if we see this as a common cause of mistake - we can update it to be an exception later

Reply:
Sounds good 👍